hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveTableUtils.java

```diff
@@ -56,7 +56,9 @@ private LegacyHiveTableUtils() {
   static Schema getSchema(org.apache.hadoop.hive.metastore.api.Table table) {
     Map<String, String> props = getTableProperties(table);
     String schemaStr = props.get("avro.schema.literal");
-    org.apache.avro.Schema avroSchema = schemaStr != null ? new org.apache.avro.Schema.Parser().parse(schemaStr) : null;
+    // Disable default value validation for backward compatibility with Avro 1.7
+    org.apache.avro.Schema avroSchema =
+        schemaStr != null ? new org.apache.avro.Schema.Parser().setValidateDefaults(false).parse(schemaStr) : null;
```

Comment:
I just have a general question here: I understand here we use `setValidateDefaults(false)` to bypass mal-formatted default values.
Is it the case that in the later code logic, we will only fail when the subsequent readers try to read such mal-formatted default value fields, and code runs gracefully otherwise; or is it the case that the later logic like schema conversion: eg things like avro schema -> iceberg schema conversion might still fail?
To me the correct behavior is the first, have we verified this?

Reply:
Good question! Yes, it will fail only when subsequent readers try to read such malformed default values because it is replaced with a "null" default when we skip default validations (I believe this is similar to avro 1.7 behavior). This is as opposed to avro 1.9 behavior which fails whenever reading the table, regardless if has data or not, and whether the data has the defaulted field filled or not (i.e., need to use the default value or not)

Below are different scenarios:

1) reading from empty table with malformed default: 

scala> val df = spark.sql("select * from u_sguirgui.test10")
[WARNING] Avro: Invalid default for field id: "one" not a "int"
df: org.apache.spark.sql.DataFrame = [id: int, address: string]

scala> df.show
+---+-------+
| id|address|
+---+-------+
+---+-------+

2) reading from table with malformed defaults, but rows populate all fields (does not need to  user the default value):
scala> spark.sql("select * from u_sguirgui.test10").show
21/09/21 05:09:47 WARN HiveConf: HiveConf of name hive.metastore.token.signature does not exist
Hive history file=/grid/hive_query_logs/sguirgui/hive_job_log_bd3f4c79-b990-4378-80d9-107944b1f034_1771310506.txt
[WARNING] Avro: Invalid default for field id: "one" not a "int"
21/09/21 05:09:49 WARN HiveConf: HiveConf of name hive.metastore.token.signature does not exist
21/09/21 05:09:49 WARN HiveConf: HiveConf of name hive.metastore.token.signature does not exist
+---+---------+                                                                 
| id|  address|
+---+---------+
|  2|address 2|
+---+---------+


3) reading from table with malformed defaults, and some rows has missing defaulted field:

scala> spark.sql("select * from u_sguirgui.test10").show
[WARNING] Avro: Invalid default for field id: "one" not a "int"
[Stage 1:>                                                          (0 + 1) / 1]21/09/21 05:17:26 WARN TaskSetManager: Lost task 0.0 in stage 1.0 (TID 1, ltx1-hcl9074.grid.linkedin.com, executor 2): java.lang.IllegalArgumentException: Missing required field that has no default value: expectedField: 0: id: required int, avroField: null, record: {"type":"record","name":"test11","namespace":"u_sguirgui","fields":[{"name":"address","type":["null","string"],"default":null,"field-id":1}]}
	at dali2_shaded.org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:441)
	at dali2_shaded.org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:112)
	at dali2_shaded.org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:42)
	at dali2_shaded.org.apache.iceberg.avro.AvroCustomOrderSchemaVisitor.visit(AvroCustomOrderSchemaVisitor.java:51)
	at dali2_shaded.org.apache.iceberg.avro.AvroSchemaUtil.buildAvroProjection(AvroSchemaUtil.java:106)